File: go/internal/agent/setup/parkstore_sqlite.go
package setup

import (
	"context"
	"errors"
	"fmt"
	"time"

	"github.com/glebarez/sqlite"
	"gorm.io/gorm"
	"gorm.io/gorm/logger"
)

// parkRow is the gorm model backing the sqlite ParkStore. The pr_key column doubles as
// the resume index ("" when the run is not parked); making it the column rather than a
// separate map means re-parking under a new key cannot leak a stale index entry. The
// workflow column scopes every claim to the owning engine (see ParkRecord.Workflow), so
// the index is really (workflow, pr_key).
type parkRow struct {
	SessionID string `gorm:"primaryKey"`
	Workflow  string `gorm:"index:idx_workflow_pr_key,priority:1"`
	PRKey     string `gorm:"index:idx_workflow_pr_key,priority:2"`
	CallID    string
	Attempts  int
	Params    string
	// autoUpdateTime:false is load-bearing, not decoration. `UpdatedAt` is a magic field
	// name to gorm, which otherwise overwrites it with the current time on every save —
	// silently discarding the caller's value, so a record could never be written as already
	// stale and this backend would diverge from memory/firestore. The conformance suite
	// catches it if the tag is ever dropped.
	UpdatedAt time.Time `gorm:"autoUpdateTime:false"`
}

func (parkRow) TableName() string { return "parked_runs" }

// parkRow mirrors ParkRecord field for field (only the gorm tags differ), so these are
// plain conversions. Adding a field to one and not the other is then a compile error rather
// than a field silently dropped on the way to or from storage.
func (r parkRow) toRecord() ParkRecord { return ParkRecord(r) }

func rowFromRecord(r ParkRecord) parkRow { return parkRow(r) }

// sqliteParkStore persists park records to a sqlite file so they survive a restart. It is
// the park-record counterpart of the sqlite session backend and shares its DSN/file.
type sqliteParkStore struct {
	db *gorm.DB
}

// newSQLiteParkStore opens (and migrates) a park store over the sqlite file at dsn.
func newSQLiteParkStore(dsn string) (ParkStore, error) {
	db, err := gorm.Open(sqlite.Open(dsn), &gorm.Config{
		PrepareStmt: true,
		Logger:      logger.Default.LogMode(logger.Silent),
	})
	if err != nil {
		return nil, fmt.Errorf("open sqlite park store: %w", err)
	}
	sqlDB, err := db.DB()
	if err != nil {
		return nil, fmt.Errorf("sqlite park store db handle: %w", err)
	}
	// SQLite serializes writers; a single pooled connection makes the claim CAS and
	// Put/Sweep contention-free within the process, WAL lets the separate session pool read
	// the shared file without blocking, and busy_timeout makes any cross-pool write wait
	// rather than fail with SQLITE_BUSY.
	sqlDB.SetMaxOpenConns(1)
	for _, pragma := range []string{"PRAGMA journal_mode=WAL", "PRAGMA busy_timeout=5000"} {
		if err := db.Exec(pragma).Error; err != nil {
			return nil, fmt.Errorf("%s: %w", pragma, err)
		}
	}
	if err := db.AutoMigrate(&parkRow{}); err != nil {
		return nil, fmt.Errorf("migrate parked_runs: %w", err)
	}
	return &sqliteParkStore{db: db}, nil
}

func (s *sqliteParkStore) Put(ctx context.Context, r ParkRecord) error {
	db := s.db.WithContext(ctx)
	if r.PRKey != "" {
		// One active row per (workflow, pr_key): clear it on any OTHER session still holding it,
		// so resolve/sweep have a single winner (the index is non-unique). Scoped by workflow so
		// a sibling engine parked on the same PR number is left alone.
		if err := db.Model(&parkRow{}).
			Where("workflow = ? AND pr_key = ? AND session_id <> ?", r.Workflow, r.PRKey, r.SessionID).
			Update("pr_key", "").Error; err != nil {
			return err
		}
	}
	// Upsert by primary key (session id). Save rewrites every column, so the pr_key index
	// follows the record automatically.
	row := rowFromRecord(r)
	return db.Save(&row).Error
}

func (s *sqliteParkStore) Get(ctx context.Context, sessionID string) (ParkRecord, bool, error) {
	var row parkRow
	err := s.db.WithContext(ctx).First(&row, "session_id = ?", sessionID).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return ParkRecord{}, false, nil
	}
	if err != nil {
		return ParkRecord{}, false, err
	}
	return row.toRecord(), true, nil
}

func (s *sqliteParkStore) ResolveByPRKey(ctx context.Context, workflow, prKey string) (ParkRecord, bool, error) {
	if prKey == "" {
		return ParkRecord{}, false, nil // an empty key would match unparked rows (pr_key='')
	}
	db := s.db.WithContext(ctx)
	var row parkRow
	err := db.First(&row, "workflow = ? AND pr_key = ?", workflow, prKey).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return ParkRecord{}, false, nil
	}
	if err != nil {
		return ParkRecord{}, false, err
	}
	return s.claim(db, row)
}

func (s *sqliteParkStore) Sweep(ctx context.Context, workflow string, cutoff time.Time) ([]ParkRecord, error) {
	db := s.db.WithContext(ctx)
	var rows []parkRow
	if err := db.Where("workflow = ? AND pr_key <> '' AND updated_at < ?", workflow, cutoff).Find(&rows).Error; err != nil {
		return nil, err
	}
	out := make([]ParkRecord, 0, len(rows))
	var errs []error
	for _, row := range rows {
		// Claim each candidate; a per-row error skips that row (it stays parked for the next
		// sweep) rather than discarding the rows already claimed this pass.
		rec, ok, err := s.claimStale(db, row, cutoff)
		if err != nil {
			errs = append(errs, err)
			continue
		}
		if ok {
			rec.PRKey = row.PRKey // restore for the caller (timeout sweep needs the PR)
			out = append(out, rec)
		}
	}
	return out, errors.Join(errs...)
}

// claim is the atomic compare-and-swap that backs ResolveByPRKey: a conditional UPDATE
// clears pr_key only while it is still set, so of N concurrent claimers exactly one (the
// writer SQLite lets through first) gets RowsAffected==1; the rest see 0 and no-op. The
// per-run row is retained (only pr_key is cleared) so a retry can still read its params.
// The row was already selected by (workflow, pr_key), and session_id is the primary key,
// so this CAS is inherently workflow-scoped.
func (s *sqliteParkStore) claim(db *gorm.DB, row parkRow) (ParkRecord, bool, error) {
	return execClaim(db.Model(&parkRow{}).
		Where("session_id = ? AND pr_key = ?", row.SessionID, row.PRKey), row)
}

// claimStale is the sweep's CAS: like claim, but also requires parked_at < cutoff, so a
// row that was resolved and re-parked (fresh) after the scan is left alone rather than
// cleared with a false timeout.
func (s *sqliteParkStore) claimStale(db *gorm.DB, row parkRow, cutoff time.Time) (ParkRecord, bool, error) {
	return execClaim(db.Model(&parkRow{}).
		Where("session_id = ? AND pr_key = ? AND updated_at < ?", row.SessionID, row.PRKey, cutoff), row)
}

func execClaim(q *gorm.DB, row parkRow) (ParkRecord, bool, error) {
	res := q.Update("pr_key", "")
	if res.Error != nil {
		return ParkRecord{}, false, res.Error
	}
	if res.RowsAffected == 0 {
		return ParkRecord{}, false, nil
	}
	row.PRKey = ""
	return row.toRecord(), true, nil
}

func (s *sqliteParkStore) Delete(ctx context.Context, sessionID string) error {
	return s.db.WithContext(ctx).Delete(&parkRow{}, "session_id = ?", sessionID).Error
}

func (s *sqliteParkStore) SweepOrphans(ctx context.Context, workflow string, cutoff time.Time) ([]ParkRecord, error) {
	var rows []parkRow
	if err := s.db.WithContext(ctx).
		Where("workflow = ? AND pr_key = '' AND updated_at < ?", workflow, cutoff).
		Find(&rows).Error; err != nil {
		return nil, err
	}
	out := make([]ParkRecord, 0, len(rows))
	for _, row := range rows {
		out = append(out, row.toRecord())
	}
	return out, nil
}

func (s *sqliteParkStore) ParkedCount(ctx context.Context, workflow string) (int, error) {
	var n int64
	if err := s.db.WithContext(ctx).Model(&parkRow{}).
		Where("workflow = ? AND pr_key <> ''", workflow).Count(&n).Error; err != nil {
		return 0, err
	}
	return int(n), nil
}

File: kotlin/src/main/kotlin/com/automation/agent/agent/setup/ParkStoreSqlite.kt
/*
 * The sqlite park store: the file-backed [ParkStore] for SESSION_BACKEND=sqlite, so a parked run
 * survives a restart. It shares its database file with the sqlite session service.
 *
 * adk-kotlin ships no database session service (unlike adk-go/adk-js), so this is hand-rolled on
 * raw JDBC. SQLite serializes writers, so a single pooled connection guarded by a mutex makes the
 * claim CAS and put/sweep contention-free within the process; WAL lets the session service read the
 * shared file without blocking, and busy_timeout makes any cross-connection write wait rather than
 * fail with SQLITE_BUSY. The `pr_key` column doubles as the resume index ('' when not parked), so
 * re-parking under a new key cannot leak a stale index entry.
 */
package com.automation.agent.agent.setup

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.sql.Connection
import java.sql.DriverManager
import java.time.Instant

/** A durable park store backed by a local SQLite file via raw JDBC. */
class SqliteParkStore(dsn: String) : ParkStore {
    private val mutex = Mutex()
    private val conn: Connection = DriverManager.getConnection("jdbc:sqlite:$dsn").apply {
        autoCommit = true
        createStatement().use { st ->
            st.execute("PRAGMA journal_mode=WAL")
            st.execute("PRAGMA busy_timeout=5000")
            st.execute(
                """
                CREATE TABLE IF NOT EXISTS parked_runs (
                    session_id TEXT PRIMARY KEY,
                    pr_key     TEXT NOT NULL DEFAULT '',
                    call_id    TEXT NOT NULL DEFAULT '',
                    attempts   INTEGER NOT NULL DEFAULT 0,
                    params     TEXT NOT NULL DEFAULT '',
                    parked_at  INTEGER
                )
                """.trimIndent(),
            )
            st.execute("CREATE INDEX IF NOT EXISTS idx_parked_runs_pr_key ON parked_runs(pr_key)")
        }
    }

    private suspend fun <T> tx(block: (Connection) -> T): T =
        mutex.withLock { withContext(Dispatchers.IO) { block(conn) } }

    // Like [tx], but runs the block in a real transaction so a multi-statement write either lands
    // whole or not at all (no partial park-record state on a mid-op failure). The shared connection
    // is mutex-serialized, so toggling autoCommit here cannot race another caller.
    private suspend fun <T> txn(block: (Connection) -> T): T =
        mutex.withLock {
            withContext(Dispatchers.IO) {
                conn.autoCommit = false
                try {
                    val result = block(conn)
                    conn.commit()
                    result
                } catch (e: Throwable) {
                    runCatching { conn.rollback() }
                    throw e
                } finally {
                    conn.autoCommit = true
                }
            }
        }

    override suspend fun put(record: ParkRecord) = txn { c ->
        if (record.prKey.isNotEmpty()) {
            // One active row per pr_key: clear it on any OTHER session still holding it, so
            // resolve/sweep have a single winner.
            c.prepareStatement("UPDATE parked_runs SET pr_key = '' WHERE pr_key = ? AND session_id <> ?").use { ps ->
                ps.setString(1, record.prKey)
                ps.setString(2, record.sessionId)
                ps.executeUpdate()
            }
        }
        // Upsert by primary key (session id): rewrites every column so the pr_key index follows.
        c.prepareStatement(
            """
            INSERT INTO parked_runs (session_id, pr_key, call_id, attempts, params, parked_at)
            VALUES (?, ?, ?, ?, ?, ?)
            ON CONFLICT(session_id) DO UPDATE SET
                pr_key = excluded.pr_key, call_id = excluded.call_id, attempts = excluded.attempts,
                params = excluded.params, parked_at = excluded.parked_at
            """.trimIndent(),
        ).use { ps ->
            ps.setString(1, record.sessionId)
            ps.setString(2, record.prKey)
            ps.setString(3, record.callId)
            ps.setInt(4, record.attempts)
            ps.setString(5, record.params)
            setNullableMillis(ps, 6, record.parkedAt)
            ps.executeUpdate()
        }
        Unit
    }

    override suspend fun get(sessionId: String): ParkRecord? = tx { c ->
        c.prepareStatement("SELECT session_id, pr_key, call_id, attempts, params, parked_at FROM parked_runs WHERE session_id = ?").use { ps ->
            ps.setString(1, sessionId)
            ps.executeQuery().use { rs -> if (rs.next()) rowToRecord(rs) else null }
        }
    }

    override suspend fun resolveByPrKey(prKey: String): ParkRecord? = txn { c ->
        if (prKey.isEmpty()) return@txn null // an empty key would match unparked rows (pr_key='')
        val row = c.prepareStatement("SELECT session_id, pr_key, call_id, attempts, params, parked_at FROM parked_runs WHERE pr_key = ? LIMIT 1").use { ps ->
            ps.setString(1, prKey)
            ps.executeQuery().use { rs -> if (rs.next()) rowToRecord(rs) else null }
        } ?: return@txn null
        // CAS: clear pr_key only while it is still set, so of N concurrent claimers exactly one
        // gets a row updated; the rest see 0 and no-op. The row is retained so a retry reads params.
        val claimed = c.prepareStatement("UPDATE parked_runs SET pr_key = '' WHERE session_id = ? AND pr_key = ?").use { ps ->
            ps.setString(1, row.sessionId)
            ps.setString(2, prKey)
            ps.executeUpdate()
        }
        if (claimed == 1) row.copy(prKey = prKey) else null
    }

    override suspend fun delete(sessionId: String) = tx { c ->
        c.prepareStatement("DELETE FROM parked_runs WHERE session_id = ?").use { ps ->
            ps.setString(1, sessionId)
            ps.executeUpdate()
        }
        Unit
    }

    override suspend fun sweep(cutoff: Instant): List<ParkRecord> = txn { c ->
        val cutoffMs = cutoff.toEpochMilli()
        val candidates = c.prepareStatement("SELECT session_id, pr_key, call_id, attempts, params, parked_at FROM parked_runs WHERE pr_key <> '' AND parked_at < ?").use { ps ->
            ps.setLong(1, cutoffMs)
            ps.executeQuery().use { rs -> buildList { while (rs.next()) add(rowToRecord(rs)) } }
        }
        buildList {
            for (row in candidates) {
                // claimStale CAS: also require parked_at < cutoff so a row resolved + re-parked
                // (fresh) after the scan is left alone rather than cleared with a false timeout.
                val claimed = c.prepareStatement("UPDATE parked_runs SET pr_key = '' WHERE session_id = ? AND pr_key = ? AND parked_at < ?").use { ps ->
                    ps.setString(1, row.sessionId)
                    ps.setString(2, row.prKey)
                    ps.setLong(3, cutoffMs)
                    ps.executeUpdate()
                }
                if (claimed == 1) add(row) // row.prKey kept for the caller (timeout sweep needs the PR)
            }
        }
    }

    override suspend fun parkedCount(): Int = tx { c ->
        c.prepareStatement("SELECT COUNT(*) FROM parked_runs WHERE pr_key <> ''").use { ps ->
            ps.executeQuery().use { rs -> if (rs.next()) rs.getInt(1) else 0 }
        }
    }

    override fun close() {
        conn.close()
    }

    private fun rowToRecord(rs: java.sql.ResultSet): ParkRecord {
        val parkedMs = rs.getLong("parked_at")
        val parkedAt = if (rs.wasNull()) null else Instant.ofEpochMilli(parkedMs)
        return ParkRecord(
            sessionId = rs.getString("session_id"),
            prKey = rs.getString("pr_key"),
            callId = rs.getString("call_id"),
            attempts = rs.getInt("attempts"),
            params = rs.getString("params"),
            parkedAt = parkedAt,
        )
    }

    private fun setNullableMillis(ps: java.sql.PreparedStatement, idx: Int, instant: Instant?) {
        if (instant == null) ps.setNull(idx, java.sql.Types.INTEGER) else ps.setLong(idx, instant.toEpochMilli())
    }
}
